### PR TITLE
Implements Bulk processing dictionary API

### PR DESCRIPTION
**Pull request by Barbayar**
## Description
Implements [Bulk processing dictionary API](https://github.com/DataDog/zstd/blob/718c6ee5063f765c9bd64a10de17968e426b9c0c/zstd.h#L841).

## Bulk processing dictionary API
It allows you to compress files with a pre-trained dictionary. Because, the compress file doesn't contain a dictionary, you can save some bytes. The save is significant if your files are smaller and similar.

P.S. With [Streaming API](https://github.com/DataDog/zstd/blob/1.x/zstd_stream.go), we can compress/decompress files with a pre-trained dictionary. But, in order to compress/decompress multiple files, we need to re-create writer/reader every time we compress/decompress. It's very inefficient, because it needs to load the dictionary every time.

## Usage
```
p, err := NewBulkProcessor(dict, BestSpeed)
if err != nil {
	fmt.Fatalf("failed to create a BulkProcessor")
}

for payload := range in {
	compressed, err := p.Compress(nil, payload)
	if err != nil {
		fmt.Fatalf("failed to compress")
	}

	uncompressed, err := p.Decompress(nil, compressed)
	if err != nil {
		fmt.Fatalf("failed to decompress")
	}
}
```

### CODE REVIEW

File: zstd_bullk_test.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+	wordCount := 10 + rand.Intn(100) // 10 - 109
+	result := []string{}
+	for i := 0; i < wordCount; i++ {
+		result = append(result, words[rand.Intn(len(words))])
+	}
+
+	return strings.Join(result, " ")
+}
+
+func TestCompressAndDecompress(t *testing.T) {
```
Comment (Viq111) [inline zstd_bullk_test.go:30]: can you prefix the tests with `TestBulk` so we don't have collisions in the future ?
https://github.com/DataDog/zstd/blob/1.x/zstd_test.go#L59
(i.e: `TestBulkCompressDecompress`)

File: zstd_bullk_test.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+	var b64 = base64.StdEncoding
```
Comment (Viq111) [inline zstd_bullk_test.go:31]: nit: that var seems a bit redundant, either define it at the test package level or don't use the var ?

File: zstd_bullk_test.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	if err != nil {
+		t.Fatalf("failed to create a BulkProcessor")
+	}
+
+	for i := 0; i < 100; i++ {
+		payload := []byte(getRandomText())
+
+		compressed, err := p.Compress(nil, payload)
+		if err != nil {
+			t.Fatalf("failed to compress")
```
Comment (Viq111) [inline zstd_bullk_test.go:47]: those are probably `t.Error` not `t.Fatal`.
and you can probably rename all `Fatalf` to `Fatal`

File: zstd_bullk_test.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	payload, err := p.Compress(nil, []byte("We're building a platform that engineers love to use. Join us, and help usher in the future."))
+	if err != nil {
+		b.Fatalf("failed to compress")
+	}
+	for n := 0; n < b.N; n++ {
+		_, err := p.Decompress(nil, payload)
+		if err != nil {
+			b.Fatalf("failed to decompress")
+		}
+		b.SetBytes(int64(len(payload)))
```
Comment (Viq111) [inline zstd_bullk_test.go:148]: this should probably be outside the loop

File: zstd_bullk_test.go [hunk 5]
```diff
@@ -0,0 +1,152 @@
+package zstd
```
Comment (Viq111) [inline zstd_bullk_test.go:1]: could you maybe test corner cases (see other tests for `zstd_test.go` and `zstd_stream_test.go` and `zstd_ctx_test.go`) ? Also I think there is some general cleanup you could do to make the code clearer (I pointed some below)

**Review by Viq111 — COMMENTED** [review]
Thanks for the PR! The structure looks straight forward, I've added a couple of comments to make it more clear and robust for an external audience

File: zstd_bulk.go [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+// Compress compresses the `src` with the dictionary
+func (p *BulkProcessor) Compress(dst, src []byte) ([]byte, error) {
+	bound := CompressBound(len(src))
+	if cap(dst) >= bound {
+		dst = dst[0:bound]
+	} else {
+		dst = make([]byte, bound)
+	}
+
+	var cSrc unsafe.Pointer
```
[the remark this answers is not in the document]
Comment (Barbayar) [reply zstd_bulk.go:48]: Okay, this was new to me. Thanks for pointing out.

**Review by Viq111 — APPROVED** [review]
Thanks! this looks great!
